### PR TITLE
feat: Tag binding for data points

### PR DESCRIPTION
**Pull request by danielherrerohernando**
## What’s the focus of this PR
This PR adds the ability to bind the related tags with each data point.
It simply adds two modules (using ES modules in the browser), one for the data (image dimensions and tags) and the other one providing logic to bind point coords to its semantic meaning.

Now the backend receives a point object containing a new set of fields ("level1","level2"...)
`
{
x: 131
y: 389,
level1: 'Happy',
level2: 'Proud',
level3: 'Successful'
}`

Depending on where the point is located, 4 different situations can be found:
- If the point is located at the utmost level of specificity, it will contain a set of fields with the **three** level tags defined in increasing order of specificity. (i.e. see the example above)
- If the point is located at the medium level of specificity, it will contain the **two** related levels defined in increasing order of specificity. 
`
{
x: 131
y: 389,
level1: 'Happy',
level2: 'Proud',
level3: null
}`
- If the point is located at the lower level of specificity, it will contain just level1 defined.
`
{
x: 131
y: 389,
level1: 'Happy',
level2: null,
level3: null
}` 
- If the point is not located inside the circle the tag fields will contain nulls. 
`
{
x: 131
y: 389,
level1: null,
level2: null,
level3: null
}`

## How to review this PR
Just check the module approach and the basic geometry-related logic. It's simply a linear transformation from cartesian coordinates to polar ones.

Before submitting this PR, I made sure:

- [x] The code builds clean without any errors or warnings
- [x] I'm using our guidelines

### CODE REVIEW

**Comment by feliun** [comment]
@danielherrerohernando thanks for this.

Can we explore the idea of this other format?

```
{ x: 131 y: 389, level1: 'Happy', level2: 'Proud', level3: 'Successful'] }
```

(or different labels for levels). The reason why I prefer an object-based structure is because that way we could query data straight from a DB (with arrays is way less comfortable). Thoughts?

**Comment by danielherrerohernando** [comment]
> @danielherrerohernando thanks for this.
> 
> Can we explore the idea of this other format?
> 
> ```
> { x: 131 y: 389, level1: 'Happy', level2: 'Proud', level3: 'Successful'] }
> ```
> 
> (or different labels for levels). The reason why I prefer an object-based structure is because that way we could query data straight from a DB (with arrays is way less comfortable). Thoughts?

Sure, how do you want it? 
I.e. when level3 is undefined, do you prefer an empty string or directly an undefined? or even no field present if there is no value for it? @feliun

**Comment by feliun** [comment]
@danielherrerohernando **null** as it's more db friendly.

**Comment by danielherrerohernando** [comment]
@feliun the PR has been updated with the desired format. I´ve also updated the PR description with the appropriate changes.

File: static/data/index.mjs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+export const imageDimensions = {
+  '0level': 0,
```
Comment (feliun) [inline static/data/index.mjs:2]: the variable name should be `dimensionsByLevel` and keys 0,1,2,3. A bit less clunky

File: static/scripts/geometry.mjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { imageDimensions, imageTags } from '../data/index.mjs'
+
+const numOfTags = Object.keys(imageTags).length;
+
+const getDistance = (horizontal,vertical) => ((horizontal**2+vertical**2)**(1/2));
+
+const calcArcsin = (a1,a2) => Math.asin(Math.abs(a1)/getDistance(a1,a2));
+
+const getAngle = (horizontal,vertical) => {
+  if(horizontal>=0 && vertical>=0) return calcArcsin(horizontal,vertical)
```
Comment (feliun) [inline static/scripts/geometry.mjs:10]: can we evaluate inside these if statement functions to abstract the reader away from the actually comparisons? Faster to read and more maintainable

File: static/scripts/geometry.mjs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  if(horizontal>=0 && vertical<0) return calcArcsin(vertical,horizontal)+Math.PI/2
+  if(horizontal<0 && vertical<0) return calcArcsin(horizontal,vertical)+Math.PI
+  if(horizontal<0 && vertical>=0) return calcArcsin(vertical,horizontal)+3/2*Math.PI
+};
+
+export const getTags = (horizontal,vertical) => {
+  const tagSlice = Math.floor(getAngle(horizontal,vertical)/(2*Math.PI/numOfTags))
+  const distance = getDistance(horizontal,vertical)
+  const tags = imageTags[tagSlice].reduce((ac,emotion,idx) => {
+    ac[`level${idx+1}`] = distance>=imageDimensions[`${idx}level`] && distance<=imageDimensions[`3level`] ? emotion : null
```
Comment (feliun) [inline static/scripts/geometry.mjs:20]: same comment as before, could you wrap these details in a function? Like ```sameLevel() ? emotion: null``` or similar

File: static/data/index.mjs (hunk 1 above)
Comment (danielherrerohernando) [reply static/data/index.mjs]: 100% agree, done.

File: static/scripts/geometry.mjs (hunk 2 above)
Comment (danielherrerohernando) [reply static/scripts/geometry.mjs]: agree, it´s already been refactored.

File: static/scripts/geometry.mjs (hunk 3 above)
Comment (danielherrerohernando) [reply static/scripts/geometry.mjs]: agree, done.

**Comment by danielherrerohernando** [comment]
nice feedback @feliun, the code has been updated according to the comments and I´ve also added eslint to provide basic linting, even for the HTML files with scripts in place

**Comment by feliun** [comment]
@danielherrerohernando if you read what happens after you send this POST, it goes against a lambda function. We also need to change a couple of things here. Otherwise the data you produced will be missed out:

- Manage to store our whole new body inside the s3 file: https://github.com/guidesmiths/mood/blob/master/lambda-backend/handler.js#L47 (I'd also extend the input object with a timestamp from the time when it's received by the backend)
- Replace this info with the data inside the retrieved object (x, y, and timestamp) https://github.com/guidesmiths/mood/blob/master/lambda-backend/handler.js#L30-L32

File: static/scripts/geometry.mjs [hunk 4]
```diff
@@ -1,24 +1,39 @@
-import { imageDimensions, imageTags } from '../data/index.mjs'
+import { dimensionsByLevel, imageTags } from '../data/index.mjs';
+
+const quadrantOffsets = {
```
Comment (feliun) [inline static/scripts/geometry.mjs:3]: nice!

**Comment by danielherrerohernando** [comment]
> @danielherrerohernando if you read what happens after you send this POST, it goes against a lambda function. We also need to change a couple of things here. Otherwise the data you produced will be missed out:
> 
> * Manage to store our whole new body inside the s3 file: https://github.com/guidesmiths/mood/blob/master/lambda-backend/handler.js#L47 (I'd also extend the input object with a timestamp from the time when it's received by the backend)
> * Replace this info with the data inside the retrieved object (x, y, and timestamp) https://github.com/guidesmiths/mood/blob/master/lambda-backend/handler.js#L30-L32

@feliun I´ve just pushed those modifications for the handler. Basically we are using the Body field where we place the point object stringified, containing all the aforementioned fields including the timestamp, generated when the lambda executes, in a human readable format: 
`{ x: 131 y: 389, level1: 'Happy', level2: 'Proud', level3: 'Successful', timestamp:''2020-04-29T11:07:32.840Z' }`
Then the lambda's get method has been updated accordingly to use the data coming in the Body field instead of the previous Key field.

**Comment by feliun** [comment]
@danielherrerohernando excellent work, thanks. I will merge and deploy this during the next days

**Pull request merged**